### PR TITLE
API_introduce_cou_level_access

### PR DESCRIPTION
**Pull request by ioigoume**
```sql
alter table cm_api_users add column identifier_type varchar(32);
alter table cm_api_users add column cou_id_list text;
```

### Get all User COU memberships knowing the user identifier( allowed for CM, CO Admins)
```bash
GET https://{{host}}/registry/co_person_roles/co/:coid/identifier/:identifier.json
Cache-Control: no-cache
Accept: application/json
Authorization: Basic {{user_test}} {{password_test}}
```
### Add User to a COU knowing the COU ID, Person Identifier
```bash
POST https://{{host}}/registry/co_person_roles.json
Cache-Control: no-cache
Content-Type: application/json; charset=utf-8
Accept: application/json
Authorization: Basic {{user_test}} {{password_test}}

{
  "RequestType": "CoPersonRoles",
  "Version": "1.0",
  "CoPersonRoles": [
    {
      "Version": "1.0",
      "Person": {
        "Type": "CO",
        "Identifier": {
          "Type": "epuid",
          "Id": "<IDENTIFIER>"
        }
      },
      "CouId": "<COU ID>",
      "Affiliation": "member",
      "Title": "",
      "Status": "Active",
      "ValidFrom": "",
      "ValidThrough": ""
    }
  ]
}
```
### Get all Users under a COU
```bash
GET https://{{host}}/registry/co_person_roles.json?couid=<cou_id>
Cache-Control: no-cache
Accept: application/json
Authorization: Basic {{user_test}} {{password_test}}
```

### Add New Role knowing CO ID, COU Name and Person Identifier
```bash
POST https://{{host}}/registry/co_person_roles.json
Cache-Control: no-cache
Content-Type: application/json; charset=utf-8
Accept: application/json
Authorization: Basic {{user_test}} {{password_test}}

{
  "RequestType": "CoPersonRoles",
  "Version": "1.0",
  "CoPersonRoles": [
    {
      "Version": "1.0",
      "Person": {
        "Type": "CO",
        "Identifier": {
          "Type": "epuid",
          "Id": "<IDENTIFIER>"
        }
      },
      "Cou": {
        "Name": "<VO NAME>",
        "CoId": "<CO ID>"      
      },
      "Affiliation": "member",
      "Title": "Engineer",
      "Status": "Active",
      "ValidFrom": "",
      "ValidThrough": ""
    }
  ]
}
```

### CODE REVIEW

File: app/View/CoPersonRoles/json/retrieve.ctp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+              $a[$k] = $m[$req][$k];
+              break;
+          }
+        }
+      }
+    }
+
+    if(($req == 'CoPerson' || $req == 'OrgIdentity')
+      && isset($m['Name']))
+    {
```
Comment (lionick) [inline app/View/CoPersonRoles/json/retrieve.ctp]: Fix the position of curly braces

File: app/View/CoPersonRoles/json/retrieve.ctp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+ */
+
+// Get a pointer to our model
+$model = $this->name;                     // Unlike in the controllers, this is CamelPlural
+$req = Inflector::singularize($model);
+$modelpl = Inflector::tableize($req);
+$modelcc = $model;
+
+if(isset($$modelpl))
+{
```
Comment (lionick) [inline app/View/CoPersonRoles/json/retrieve.ctp]: Fix the position of curly braces

File: app/View/CoPersonRoles/json/retrieve.ctp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  $ms = array();
+
+  foreach($$modelpl as $m)
+  {
```
Comment (lionick) [inline app/View/CoPersonRoles/json/retrieve.ctp]: Fix the position of curly braces

File: app/Controller/ApiUsersController.php [hunk 4]
```diff
@@ -63,6 +63,22 @@ function beforeFilter() {
...
+   *
+   * @since  COmanage Registry v3.4.0 RCIAM
+   */
+
+  function beforeRender() {
+    parent::beforeRender();
+
+    if (!$this->request->is('restful')) {
+      $this->set('vv_cous', $this->ApiUser->Co->Cou->allCous($this->cur_co['Co']['id']));
+      // and identifier types
```
Comment (lionick) [inline app/Controller/ApiUsersController.php]: do you mean add?

**Review by lionick — CHANGES_REQUESTED** [review]

File: app/View/CoPersonRoles/json/retrieve.ctp (hunk 2 above)
Comment (ioigoume) [reply app/View/CoPersonRoles/json/retrieve.ctp]: Done

File: app/View/CoPersonRoles/json/retrieve.ctp (hunk 3 above)
Comment (ioigoume) [reply app/View/CoPersonRoles/json/retrieve.ctp]: Done

File: app/View/CoPersonRoles/json/retrieve.ctp (hunk 1 above)
Comment (ioigoume) [reply app/View/CoPersonRoles/json/retrieve.ctp]: Done

File: app/Controller/ApiUsersController.php (hunk 4 above)
Comment (ioigoume) [reply app/Controller/ApiUsersController.php]: Done

**Review by lionick — APPROVED** [review]
LGTM

**Review by NicolasLiampotis — APPROVED** [review]
LGTM

**Pull request merged**